packages/back-end/src/services/clickhouse.ts

```diff
@@ -0,0 +1,122 @@
+import * as crypto from "crypto";
+import { createClient as createClickhouseClient } from "@clickhouse/client";
+import generator from "generate-password";
+import {
+  CLICKHOUSE_HOST,
+  CLICKHOUSE_ADMIN_USER,
+  CLICKHOUSE_ADMIN_PASSWORD,
+  CLICKHOUSE_DATABASE,
+  CLICKHOUSE_MAIN_TABLE,
+  ENVIRONMENT,
+} from "back-end/src/util/secrets";
+import { DataSourceParams } from "back-end/types/datasource";
+import { ReqContext } from "back-end/types/organization";
+import { logger } from "back-end/src/util/logger";
+
+export async function createClickhouseUser(
+  context: ReqContext,
+  datasourceId: string
+): Promise<DataSourceParams> {
+  // Temporary protection to prevent users from manually changing feature flag to create Clickhouse users.
+  if (ENVIRONMENT === "production" && context.org.id != "org_24yyifrkf649iz6") {
+    throw new Error(
+      "Clickhouse user creation is only allowed for the Growthbook organization"
+    );
+  }
+
+  const client = createClickhouseClient({
+    host: CLICKHOUSE_HOST,
+    username: CLICKHOUSE_ADMIN_USER,
+    password: CLICKHOUSE_ADMIN_PASSWORD,
+    database: CLICKHOUSE_DATABASE,
+    application: "GrowthBook",
+    request_timeout: 3620_000,
+    clickhouse_settings: {
+      max_execution_time: 3600,
+    },
+  });
+
+  const orgId = context.org.id;
+  const user =
+    ENVIRONMENT === "production"
+      ? `${orgId}_${datasourceId}`
+      : `test_${orgId}_${datasourceId}`;
+  const password = generator.generate({
+    length: 30,
+    numbers: true,
+  });
+  const hashedPassword = crypto
+    .createHash("sha256")
+    .update(password)
+    .digest("hex");
+
+  const database = user;
+  const viewName = `${database}.events`;
+
+  logger.info(`creating Clickhouse database ${database}`);
+  await client.command({
+    query: `CREATE DATABASE ${database}`,
+  });
+
+  logger.info(`Creating Clickhouse user ${user}`);
+  await client.command({
+    query: `CREATE USER ${user} IDENTIFIED WITH sha256_hash BY '${hashedPassword}'`,
+  });
+
+  logger.info(`Creating Clickhouse view ${viewName}`);
+  await client.command({
+    query: `CREATE VIEW ${viewName} DEFINER=CURRENT_USER SQL SECURITY DEFINER AS SELECT * FROM ${CLICKHOUSE_MAIN_TABLE} WHERE organization = '${orgId}'`,
+  });
+
+  logger.info(`Granting select permissions on ${viewName} to ${user}`);
+  await client.command({ query: `GRANT SELECT ON ${viewName} TO ${user}` });
+
+  logger.info(`Clickhouse user ${user} created`);
+
+  const parts = CLICKHOUSE_HOST.split(":");
+  const port = parseInt(CLICKHOUSE_HOST.split(":").pop() || "9000");
+  const url = parts.join(":");
+
+  const params = {
+    port: port,
+    url: url,
+    user: user,
+    password: password,
+    database: database,
+  };
+
+  return params;
+}
+
+export async function deleteClickhouseUser(
+  datasourceId: string,
+  organization: string
+) {
+  const client = createClickhouseClient({
```

Comment:
Same protection as above? Maybe factor it out and re-use in both calls?

Reply:
That's what I'd thought at first but this is only called by `DataSourceModel#deleteDatasource` with
```
  if (datasource.type === "growthbook_clickhouse") {
    await deleteClickhouseUser(datasource.id, organization);
  }
```
so it seems safe to leave open like this. Users won't have clickhouse datasources and if somehow they do then they should be able to delete them

Reply:
I think that it would be more clear centralize the checks in the same file. Otherwise we risk missing one and/or making the code harder to proofread.

Reply:
I agree in general, but I'm thinking this flow doesn't need a check at all. We shouldn't let users create the datasources (hence the other check), but if we mess up and they're able to create one at some point, we should let them delete it (no check here).

Reply:
My concern when reading this code is: "This code is supposed to only apply to our own internal consumption at the moment, how can I make sure that this function is not executed by people from other orgs"?

Having the check right here makes it more clear. Also, when we open-up the functionality to other orgs, it would be easier to change the logic.

Reply:
I would lean towards leaving this open and just gating the creation side of things.